postalicious/public/app.js

```diff
@@ -0,0 +1,102 @@
+const methodInput = document.getElementById('method')
+const uriInput = document.getElementById('uri')
+const hostInput = document.getElementById('host')
+
+const queryKeys = document.getElementsByClassName('queryKey')
+const queryValues = document.getElementsByClassName('queryValue')
+
+const headerKeys = document.getElementsByClassName('headerKey')
+const headerValues = document.getElementsByClassName('headerValue')
+
+const bodyInput = document.getElementById('body')
+
+const rawRequestDiv = document.getElementById('rawRequest')
+const rawResponseDiv = document.getElementById('rawResponse')
+
+const methods = [
+  'GET',
+  'POST'
+]
+
+function build () {
+  if (!methods.find(method => method === methodInput.value)) {
+    alert(`You must enter one of these methods: ${methods.join(', ')}`)
+    return
+  }
+
+  if (!hostInput.value) {
+    alert('You must enter a host')
+    return
+  }
+
+  if (!uriInput.value) {
+    alert('You must enter a URI')
+    return
+  }
+
+  rawResponseDiv.textContent = ''
+
+  let request = `${methodInput.value} ${uriInput.value}`
+
+  if (methodInput.value === 'GET') {
+    const queryKeysArray = Array.from(queryKeys).map(input => input.value)
+    const queryValuesArray = Array.from(queryValues).map(input => input.value)
+
+    const queries = queryKeysArray.reduce((queries, key, index) => {
+      if (key.length !== 0 && queryValuesArray[index].length !== 0) {
+        queries.push(`${key}=${queryValuesArray[index]}`)
+      }
+      return queries
+    }, [])
+
+    if (queries.length !== 0) {
+      request += `?${queries.join('&')}`
+    }
+  }
+
+  request += ` HTTP/1.1\r\nHost: ${hostInput.value}\r\n`
+
+  const headerKeysArray = Array.from(headerKeys).map(input => input.value)
+  const headerValuesArray = Array.from(headerValues).map(input => input.value)
+
+  const headers = headerKeysArray.reduce((headers, key, index) => {
+    if (key.length !== 0 && headerValuesArray[index].length !== 0) {
+      headers.push(`${key}: ${headerValuesArray[index]}`)
+    }
+    return headers
+  }, [])
+
+  if (bodyInput.value.length !== 0) {
+    headers.push(`Content-Length: ${bodyInput.value.length}`)
+    headers.push('Content-Type: text/plain')
+  }
+
+  if (headers.length !== 0) {
```

Comment:
I'm not sure what you mean.  Which lines are you referring to?

On lines 62-67 I start to build `headers` from things that the user put in the form.

On lines 69-72 I add headers that are only needed if there is a body.

On lines 74-76 I add all these headers to the request string.